libyara/ahocorasick.c

```diff
@@ -791,11 +1557,27 @@ int yr_ac_add_string(
 
     for (int i = 0; i < atom->atom.length; i++)
     {
-      YR_AC_STATE* next_state = _yr_ac_next_state(state, atom->atom.bytes[i]);
+      if (atom->atom.mask[i] == YR_ATOM_TYPE_CLASS)
+      {
+        // for a case like [a] - the class is a literal
+        int literal = _yr_ac_class_is_literal(atom->atom.bitmap[i]);
+        if (literal >= 0)
+        {
+          atom->atom.bytes[i] = literal;
```

Comment:
With this change function `yr_ac_add_string` have the side effect of modifying the atoms that are passed to it in the list of atoms. This makes the logic harder to follow, because `yr_ac_add_string` is expected to modify the Aho-Corasick automaton, but not its input arguments like the string passed as `YR_STRING*` or the atoms passed as `YR_ATOM_LIST_ITEM*`.

I would say that this transformation from atoms of type `YR_ATOM_TYPE_CLASS` to `YR_ATOM_TYPE_LITERAL` should be done outside this function, at the place where the atoms are generated.

Reply:
Updated.